### PR TITLE
US040 Login System

### PR DESCRIPTION
**Pull request by cormac-larkin-kainos**
Added JWT Token Generator and Validator classes. Also updated unit and integration tests

### CODE REVIEW

**Review by JonnyEvans99 — APPROVED** [review]

File: src/main/java/org/kainos/ea/FloorIsJavaApplication.java [hunk 1]
```diff
@@ -30,8 +36,9 @@ protected SwaggerBundleConfiguration getSwaggerBundleConfiguration(FloorIsJavaCo
 
     @Override
     public void run(final FloorIsJavaConfiguration configuration,
-                    final Environment environment) {
+                    final Environment environment) throws NoSuchAlgorithmException {
```
Comment (shaunganley) [inline src/main/java/org/kainos/ea/FloorIsJavaApplication.java:39]: don't throw this exception from here, it should be caught and returned as a 500 in the controller

File: src/main/java/org/kainos/ea/auth/JwtGenerator.java [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+
+public class JwtGenerator {
+
+    public static final KeyPair KEYPAIR;
+
+    static {
+        try {
+            KEYPAIR = KeyPairGenerator.getInstance("RSA").generateKeyPair();
+        } catch (NoSuchAlgorithmException e) {
+            throw new RuntimeException(e);
```
Comment (shaunganley) [inline src/main/java/org/kainos/ea/auth/JwtGenerator.java]: shouldn't throw a runtime exception, can catch NoSuchAlgorithmException and return a 500 from controller

File: src/main/java/org/kainos/ea/auth/JwtGenerator.java [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+        }
+    }
```
Comment (shaunganley) [inline src/main/java/org/kainos/ea/auth/JwtGenerator.java]: can you move this to a constructor?

File: src/main/java/org/kainos/ea/auth/JwtGenerator.java [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+     * @param username
+     * @return The JWT token as a String
+     */
+    public String generateJwt(String username) {
+
+        Builder tokenBuilder = JWT.create()
+                .withIssuer("https://www.floor-is-java.com") //placeholder url
+                .withSubject(username)
+                .withClaim("jti", UUID.randomUUID().toString())
+                .withExpiresAt(Date.from(Instant.now().plusSeconds(1800))) // Token expires after 30 minutes
```
Comment (shaunganley) [inline src/main/java/org/kainos/ea/auth/JwtGenerator.java]: could you make this 8 hours?

File: src/test/java/org/kainos/ea/integration/AuthControllerIntegrationTest.java [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+@ExtendWith(DropwizardExtensionsSupport.class)
+public class AuthControllerIntegrationTest {
+
+    static final DropwizardAppExtension<FloorIsJavaConfiguration> APP = new DropwizardAppExtension<>(
+            FloorIsJavaApplication.class, null,
+            new ResourceConfigurationSourceProvider()
+    );
+    @Test
+    void loginWithValidLogin_ShouldReturnLogin() {
+        Login testLogin = new Login("admin", "admin");
```
Comment (shaunganley) [inline src/test/java/org/kainos/ea/integration/AuthControllerIntegrationTest.java]: don't hardcode valid creds, pass in as env vars

**Review by shaunganley — CHANGES_REQUESTED** [review]

File: src/main/java/org/kainos/ea/auth/JwtGenerator.java (hunk 4 above)
Comment (cormac-larkin-kainos) [reply src/main/java/org/kainos/ea/auth/JwtGenerator.java]: Updated to set token expiry to 8 hours

File: src/main/java/org/kainos/ea/auth/JwtGenerator.java (hunk 3 above)
Comment (cormac-larkin-kainos) [reply src/main/java/org/kainos/ea/auth/JwtGenerator.java]: Updated to fix this

File: src/main/java/org/kainos/ea/auth/JwtGenerator.java (hunk 2 above)
Comment (cormac-larkin-kainos) [reply src/main/java/org/kainos/ea/auth/JwtGenerator.java]: Fixed this

File: src/main/java/org/kainos/ea/FloorIsJavaApplication.java (hunk 1 above)
Comment (cormac-larkin-kainos) [reply src/main/java/org/kainos/ea/FloorIsJavaApplication.java:39]: Left this as is after discussion

File: src/test/java/org/kainos/ea/integration/AuthControllerIntegrationTest.java (hunk 5 above)
Comment (cormac-larkin-kainos) [reply src/test/java/org/kainos/ea/integration/AuthControllerIntegrationTest.java]: Updated to use env vars. Also updated README to reflect these changes

**Review by shaunganley — APPROVED** [review]

**Review by dianemorrison — APPROVED** [review]

File: src/main/java/org/kainos/ea/auth/JwtGenerator.java [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+
+    /**
+     * Generates a JWT token for a specified user
+     *
+     * @param username
+     * @return The JWT token as a String
+     */
+    public String generateJwt(String username) {
+
+        Builder tokenBuilder = JWT.create()
```
Comment (shaunganley) [inline src/main/java/org/kainos/ea/auth/JwtGenerator.java:36]: you can add the role to the token either now or as part of the auth ticket

**Review by shaunganley — APPROVED** [review]
could do with some unit tests on JwtGenerator and JwtValidator classes

File: src/main/java/org/kainos/ea/auth/JwtGenerator.java (hunk 6 above)
Comment (eoinmcbennett) [reply src/main/java/org/kainos/ea/auth/JwtGenerator.java:36]: Already in progress in the auth ticket.

**Pull request merged**